File: medium/cpp/1490.cpp
// https://leetcode.com/problems/clone-n-ary-tree/description/

/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> children;

    Node() {}

    Node(int _val) {
        val = _val;
    }

    Node(int _val, vector<Node*> _children) {
        val = _val;
        children = _children;
    }
};
*/

class Solution
{
public:
    Node* cloneTree(Node* root)
    {
        if (root == nullptr)
        {
            return nullptr;
        }

        auto cloned = new Node(root->val);

        for (auto to : root->children)
        {
            cloned->children.push_back(cloneTree(to));
        }

        return cloned;
    }
};

File: medium/python/1490.py
# https://leetcode.com/problems/clone-n-ary-tree/description/

"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children if children is not None else []
"""

class Solution:
    def cloneTree(self, root: 'Node') -> 'Node':
        return self.clone(root)

    def clone(self, root):
        if not root:
            return None

        cloned = Node(root.val)

        if root.children:
            cloned.children = []

            for child in root.children:
                cloned.children.append(self.clone(child))

        return cloned